File: tools/test_golden.zig
const std = @import("std");
const flux = @import("../src/zig/flux_constraint.zig");

const Vector = struct {
    id: i32,
    value: i32,
    constraints: []const struct { lo: i32, hi: i32 },
    expected: struct {
        error_mask: i32,
        violated_lo: i32,
        violated_hi: i32,
        violated_count: i32,
        passed: bool,
    },
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const file = try std.fs.cwd().openFile("tools/golden_vectors.json", .{});
    defer file.close();
    const data = try file.readToEndAlloc(allocator, 50 * 1024 * 1024);
    defer allocator.free(data);

    const parsed = try std.json.parseFromSlice([]const Vector, allocator, data, .{
        .ignore_unknown_fields = true,
    });
    defer parsed.deinit();
    const vectors = parsed.value;

    var mismatches: i32 = 0;
    for (vectors) |v| {
        var cs: [8]flux.Constraint = undefined;
        for (v.constraints, 0..) |c, i| {
            cs[i] = .{ .lo = c.lo, .hi = c.hi };
        }
        const checker = flux.FluxChecker{ .constraints = cs[0..v.constraints.len] };
        const r = checker.check(v.value);
        if (r.error_mask != v.expected.error_mask or r.passed != v.expected.passed) {
            mismatches += 1;
        }
    }

    const stdout = std.io.getStdOut().writer();
    try stdout.print("\nZig: {} vectors, {} mismatches\n", .{ vectors.len, mismatches });
    if (mismatches > 0) std.process.exit(1);
}

File: tools/test_golden.go
//go:build ignore
// +build ignore

package main

import (
	"encoding/json"
	"fmt"
	"os"
)

func saturate(v int) int {
	if v < -127 { return -127 }
	if v > 127 { return 127 }
	return v
}

type Constraint struct { Lo, Hi int }
type Vector struct {
	Id         int
	Value      int
	Constraints []Constraint
	Expected   struct {
		ErrorMask     int `json:"error_mask"`
		ViolatedLo    int `json:"violated_lo"`
		ViolatedHi    int `json:"violated_hi"`
		ViolatedCount int `json:"violated_count"`
		Passed        bool
	}
}

func check(cs []Constraint, value int) (errorMask, violatedLo, violatedHi, violatedCount int, passed bool) {
	v := saturate(value)
	for i, c := range cs {
		lo, hi := saturate(c.Lo), saturate(c.Hi)
		loFail := v < lo
		hiFail := v > hi
		if loFail || hiFail { errorMask |= (1 << i); violatedCount++ }
		if loFail { violatedLo |= (1 << i) }
		if hiFail { violatedHi |= (1 << i) }
	}
	passed = errorMask == 0
	return
}

func main() {
	data, err := os.ReadFile("tools/golden_vectors.json")
	if err != nil { fmt.Println("Error:", err); os.Exit(1) }
	
	var vectors []Vector
	json.Unmarshal(data, &vectors)
	
	mismatches := 0
	for _, v := range vectors {
		em, vlo, vhi, vc, passed := check(v.Constraints, v.Value)
		_ = vlo; _ = vhi
		if em != v.Expected.ErrorMask || passed != v.Expected.Passed || vc != v.Expected.ViolatedCount {
			mismatches++
			if mismatches <= 5 {
				fmt.Printf("MISMATCH #%d: value=%d got mask=%d passed=%v expected mask=%d\n",
					v.Id, v.Value, em, passed, v.Expected.ErrorMask)
			}
		}
	}
	
	fmt.Printf("\nGo: %d vectors, %d mismatches\n", len(vectors), mismatches)
	if mismatches > 0 { os.Exit(1) }
}
